SAML2 Metadata UI parsing exception

It seems there is a bug about Metadata UI info parsing.

Shortly, in SamlMetadataUIInfo, the methos getDisplayNames return a Collection of String by using getStringValues method explicitly.

Unfortunately, the method to retrieve the localized display name getDisplayName is leveraging on getLocalizedValues method which is expecting a list of LocalizedName objects (not strings).

Thus, the parsing is failing reporting the exception reported below.. 

- CAS server version 5.2.0-RC3-SNAPSHOT
- Container Tomcat on Linux
- Steps:
  configure SAML2 Authentication
  add metadata UI (mdui:UIInfo) into a SP metadata
  try out with a SP Initiated login

- The process ends on the authentication page with the following exception


Caused by: java.lang.ClassCastException: java.lang.String cannot be cast to org.opensaml.saml.saml2.metadata.LocalizedName
    at org.apereo.cas.support.saml.mdui.SamlMetadataUIInfo.getLocalizedValues(SamlMetadataUIInfo.java:240) ~[classes/:5.2.0-RC3-SNAPSHOT]
    at org.apereo.cas.support.saml.mdui.SamlMetadataUIInfo.getDisplayName(SamlMetadataUIInfo.java:164) ~[classes/:5.2.0-RC3-SNAPSHOT]
    at org.apereo.cas.support.saml.mdui.SamlMetadataUIInfo.getDisplayName(SamlMetadataUIInfo.java:170) ~[classes/:5.2.0-RC3-SNAPSHOT]
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_144]
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_144]
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_144]
    at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_144]
    at org.springframework.expression.spel.support.ReflectivePropertyAccessor$OptimalPropertyAccessor.read(ReflectivePropertyAccessor.java:636) ~[spring-expression-4.3.10.RELEASE.jar:4.3.10.RELEASE]
    at org.springframework.expression.spel.ast.PropertyOrFieldReference.readProperty(PropertyOrFieldReference.java:213) ~[spring-expression-4.3.10.RELEASE.jar:4.3.10.RELEASE]
    at org.springframework.expression.spel.ast.PropertyOrFieldReference.getValueInternal(PropertyOrFieldReference.java:94) ~[spring-expression-4.3.10.RELEASE.jar:4.3.10.RELEASE]
    at org.springframework.expression.spel.ast.PropertyOrFieldReference.access$000(PropertyOrFieldReference.java:46) ~[spring-expression-4.3.10.RELEASE.jar:4.3.10.RELEASE]
    at org.springframework.expression.spel.ast.PropertyOrFieldReference$AccessorLValue.getValue(PropertyOrFieldReference.java:379) ~[spring-expression-4.3.10.RELEASE.jar:4.3.10.RELEASE]
    at org.springframework.expression.spel.ast.CompoundExpression.getValueInternal(CompoundExpression.java:88) ~[spring-expression-4.3.10.RELEASE.jar:4.3.10.RELEASE]
    at org.springframework.expression.spel.ast.SpelNodeImpl.getValue(SpelNodeImpl.java:120) ~[spring-expression-4.3.10.RELEASE.jar:4.3.10.RELEASE]
    at org.springframework.expression.spel.standard.SpelExpression.getValue(SpelExpression.java:324) ~[spring-expression-4.3.10.RELEASE.jar:4.3.10.RELEASE]
    at org.thymeleaf.spring4.expression.SPELVariableExpressionEvaluator.evaluate(SPELVariableExpressionEvaluator.java:263) ~[thymeleaf-spring4-3.0.7.RELEASE.jar:3.0.7.RELEASE]

--- a/support/cas-server-support-saml-mdui-core/src/main/java/org/apereo/cas/support/saml/mdui/SamlMetadataUIInfo.java
+++ b/support/cas-server-support-saml-mdui-core/src/main/java/org/apereo/cas/support/saml/mdui/SamlMetadataUIInfo.java
@@ -143,8 +143,11 @@ public class SamlMetadataUIInfo extends DefaultRegisteredServiceUserInterfaceInf
      * @return the description
      */
     public String getDescription(final String locale) {
-        final String description = getLocalizedValues(locale, List.class.cast(this.getDescriptions()));
-        return (description != null) ? description : super.getDescription();
+        if (this.uiInfo != null) {
+            final String description = getLocalizedValues(locale, this.uiInfo.getDescriptions());
+            return (description != null) ? description : super.getDescription();
+        }
+        return super.getDescription();
     }
 
     @Override
@@ -159,8 +162,11 @@ public class SamlMetadataUIInfo extends DefaultRegisteredServiceUserInterfaceInf
      * @return the displayName
      */
     public String getDisplayName(final String locale) {
-        final String displayName = getLocalizedValues(locale, List.class.cast(this.getDisplayNames()));
-        return (displayName != null) ? displayName : super.getDisplayName();
+        if (this.uiInfo != null) {
+            final String displayName = getLocalizedValues(locale, this.uiInfo.getDisplayNames());
+            return (displayName != null) ? displayName : super.getDisplayName();
+        }
+        return super.getDisplayName();
     }
 
     @Override
@@ -175,8 +181,11 @@ public class SamlMetadataUIInfo extends DefaultRegisteredServiceUserInterfaceInf
      * @return the informationURL
      */
     public String getInformationURL(final String locale) {
-        final String informationUrl = getLocalizedValues(locale, List.class.cast(this.getInformationURLs()));
-        return (informationUrl != null) ? informationUrl : super.getInformationURL();
+        if (this.uiInfo != null) {
+            final String informationUrl = getLocalizedValues(locale, this.uiInfo.getInformationURLs());
+            return (informationUrl != null) ? informationUrl : super.getInformationURL();
+        }
+        return super.getInformationURL();
     }
 
     @Override
@@ -191,8 +200,11 @@ public class SamlMetadataUIInfo extends DefaultRegisteredServiceUserInterfaceInf
      * @return the privacyStatementURL
      */
     public String getPrivacyStatementURL(final String locale) {
-        final String privacyStatementURL = getLocalizedValues(locale, List.class.cast(this.getPrivacyStatementURLs()));
-        return (privacyStatementURL != null) ? privacyStatementURL : super.getPrivacyStatementURL();
+        if (this.uiInfo != null) {
+            final String privacyStatementURL = getLocalizedValues(locale, this.uiInfo.getPrivacyStatementURLs());
+            return (privacyStatementURL != null) ? privacyStatementURL : super.getPrivacyStatementURL();
+        }
+        return super.getPrivacyStatementURL();
     }
 
     @Override

